Tests don't compile on Clang 16 due to memory safety warnings with `-Werror`

## Description
<!--
If your issue is a bugreport, this means describing what you did,
what did you want to happen and what actually did happen.

If your issue is a feature request, describe the feature and why do you
want it.
-->
I try to run CMake for this library, which I assume just builds the examples.

### Steps to reproduce
<!--
This is only relevant for bug reports, but if you do have one,
please provide a minimal set of steps to reproduce the problem.

Usually this means providing a small and self-contained code using doctest
and specifying compiler flags/tools used if relevant.
-->

```sh
mkdir build
cd build
cmake .. -DCMAKE_CXX_COMPILER=clang++-16
make -k
```
This gave me a bunch of errors like this one:
```sh
[  0%] Building CXX object examples/all_features/CMakeFiles/all_features.dir/main.cpp.o
In file included from /home/holyblackcat/Sandbox/2/doctest-2.4.11/examples/all_features/main.cpp:2:
In file included from /home/holyblackcat/Sandbox/2/doctest-2.4.11/scripts/development_only/doctest/doctest.h:1:
/home/holyblackcat/Sandbox/2/doctest-2.4.11/doctest/parts/doctest_fwd.h:608:47: error: unsafe buffer access [-Werror,-Wunsafe-buffer-usage]
    bool isOnStack() const noexcept { return (buf[last] & 128) == 0; }
```

### Extra information
<!--
Fill in any extra information that might be important for your issue.

If your issue is a bugreport, definitely fill out at least the following.
-->
* doctest version: **v2.4.11**
* Operating System: **Ubuntu 22.04**
* Compiler+version: **Clang 16 from https://apt.llvm.org/**

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -294,6 +294,12 @@ jobs:
           if [ "${{ matrix.compiler }}" = "gcc" ]; then
             sudo apt-get install -y g++-${{ matrix.version }} g++-${{ matrix.version }}-multilib
           elif [ "${{ matrix.compiler }}" = "clang" ]; then
+            # clang-17 does not automatically provide the i386 sanitizer runtimes,
+            # but libclang-rt-V-dev:i386 does not always exist. This is the best intermediate option
+            if [ "${{ matrix.version }}" = "17" ]; then
+              sudo apt-get install -y libclang-rt-${{ matrix.version }}-dev:i386
+            fi
+
             sudo apt-get install -y clang-${{ matrix.version }} libclang-${{ matrix.version }}-dev llvm-${{ matrix.version }}-dev g++-multilib
           fi
 

--- a/doctest/doctest.h
+++ b/doctest/doctest.h
@@ -162,11 +162,13 @@
 #define DOCTEST_SUPPRESS_COMMON_WARNINGS_PUSH                                                      \
     DOCTEST_CLANG_SUPPRESS_WARNING_PUSH                                                            \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wunknown-pragmas")                                            \
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wunknown-warning-option")                                     \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wweak-vtables")                                               \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wpadded")                                                     \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wmissing-prototypes")                                         \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wc++98-compat")                                               \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wc++98-compat-pedantic")                                      \
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wunsafe-buffer-usage")                                        \
                                                                                                    \
     DOCTEST_GCC_SUPPRESS_WARNING_PUSH                                                              \
     DOCTEST_GCC_SUPPRESS_WARNING("-Wunknown-pragmas")                                              \

--- a/doctest/parts/doctest_fwd.h
+++ b/doctest/parts/doctest_fwd.h
@@ -159,11 +159,13 @@
 #define DOCTEST_SUPPRESS_COMMON_WARNINGS_PUSH                                                      \
     DOCTEST_CLANG_SUPPRESS_WARNING_PUSH                                                            \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wunknown-pragmas")                                            \
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wunknown-warning-option")                                     \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wweak-vtables")                                               \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wpadded")                                                     \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wmissing-prototypes")                                         \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wc++98-compat")                                               \
     DOCTEST_CLANG_SUPPRESS_WARNING("-Wc++98-compat-pedantic")                                      \
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wunsafe-buffer-usage")                                        \
                                                                                                    \
     DOCTEST_GCC_SUPPRESS_WARNING_PUSH                                                              \
     DOCTEST_GCC_SUPPRESS_WARNING("-Wunknown-pragmas")                                              \

--- a/scripts/cmake/common.cmake
+++ b/scripts/cmake/common.cmake
@@ -1,4 +1,5 @@
 include(CMakeParseArguments)
+include(CheckCXXCompilerFlag)
 
 # cache this for use inside of the function
 set(CURRENT_LIST_DIR_CACHED ${CMAKE_CURRENT_LIST_DIR})
@@ -187,6 +188,11 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
     add_compiler_flags(-Wno-c++98-compat-bind-to-temporary-copy)
     add_compiler_flags(-Wno-c++98-compat-local-type-template-args)
     add_compiler_flags(-Qunused-arguments -fcolor-diagnostics) # needed for ccache integration
+endif()
+
+check_cxx_compiler_flag(-Wunsafe-buffer-usage HAVE_UNSAFE_BUFFER_USAGE)
+if(HAVE_UNSAFE_BUFFER_USAGE)
+    add_compiler_flags(-Wno-unsafe-buffer-usage)
 endif()
 
 if(MSVC)

